### PR TITLE
feat: indexing-time benchmark on a pinned corpus (#1374 deliverable 4)

### PR DESCRIPTION
**Pull request by vitali87**
Part of #1374. Implements **deliverable 4 only** — the indexing benchmark.

## Why this one first

#1374 proposes its own phasing and puts this first, for a good reason: it is the only one of the four benchmarks needing **neither a model key nor a running stack**. It produces a README row today, in CI, with no infrastructure. Deliverables 1 (agentic A/B), 2 (query latency) and 3 (token cost) all require an LLM key or a live Memgraph/Qdrant, and remain open.

## What it measures — and what it deliberately does not

Runs the real `GraphUpdater` pipeline (the same parse, definition and call passes production uses) against the in-memory capturing ingestor the eval harness already relies on.

Database round-trips are excluded **by construction**, and that exclusion is the point rather than a shortcut: it isolates the cost this project controls from the cost of the two databases — which is exactly what the field-test review asked ("is the graph worth feeding?"). The module docstring says so explicitly so nobody misreads the number as end-to-end `cgr index` wall-clock. Deliverable 2 covers that, and its number will be larger.

## Measured result

Running against `codebase_rag/parsers`:

| Benchmark | Time | Graph | Peak RSS |
|---|---|---|---|
| Indexing — parsers, 135 files (cgr 0.0.770) | 5.3s | 2604 nodes / 12582 edges | 203 MiB |

Reproduce with:

```
uv run python -m benchmarks.bench_indexing codebase_rag/parsers --project-name cgr_parsers
```

Every row carries its corpus, file count and cgr version, because #1374 is explicit that *the credibility of the table is the reproducibility of its cells*. A timing without provenance is a number nobody can check.

## Two failure modes closed deliberately

**An empty corpus raises rather than reporting zeroes.** `0 files in 0.01s` reads like a very fast index rather than like a benchmark that measured nothing — the published-number form of the silent-smaller-than-truth shape this repo keeps hitting. Verified by mutation: removing the guard fails the test.

**The file count is FILE nodes only.** cgr emits *both* a File and a Module node per source file, and my first version counted either-or — it would have published a file count **twice the truth**. It was caught because the test asserts the fixture's real shape (`file_count == 3`) rather than merely that the field exists; a presence check would have shipped it.

## Testing

Red/green TDD: tests written first and observed failing at collection, then implementation. 4 tests covering real graph counts, wall-clock and memory being measured rather than defaulted, the empty-corpus refusal, and row provenance.

Note the README table row itself is **not** added here — populating the front-page table is better done once more than one benchmark can fill it, rather than landing a one-row table that then churns.

## Summary by CodeRabbit

* **New Features**
  * Added a reproducible indexing benchmark with Markdown and JSON output.
  * Reports corpus details, file/node/edge counts, indexing duration, and peak memory usage.
  * Supports optional project naming and rejects empty corpora.
  * Handles environments where peak memory metrics are unavailable.

* **Tests**
  * Added coverage for benchmark metrics, validation, formatted output, process isolation, and platform compatibility.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
Both findings were correct and both are fixed in d4872794. Thank you for reproducing them rather than asserting them — the second in particular would have shipped a misleading published number.

**1. Windows could not import the module.** `import resource` was unconditional at module scope, and it is Unix-only. Now guarded, with memory reported as an explicit `RSS_UNAVAILABLE` sentinel rendered `n/a` rather than `0` — a zero would read as "the run used no memory" instead of "this platform cannot measure it".

**2. The RSS figure was attributable to unrelated earlier work.** Fixed by measuring in a freshly spawned subprocess, as you suggested. Your point that subtraction cannot rescue it is what settled the approach: once the prior high-water exceeds the run peak, the difference is zero or negative.

Worth noting my own comment had already identified the problem — *"the delta would be negative whenever the peak was already set by earlier work in this process"* — and then shipped `max()`, which does not address it. Diagnosing it and fixing it turned out to be different acts.

**The effect is not cosmetic.** The same corpus now reports **596 MiB** where it previously reported **203 MiB**. The old figure was under-reporting because the parent process peak predated parser loading, so the published table would have understated indexing memory by roughly 3x.

| Indexing time — parsers, 135 files (cgr 0.0.770) | 7.6s | 2736 nodes / 12868 edges | 596 MiB |

**On the test.** My first draft asserted `peak < ballast_size` and failed against the *correct* implementation — a child that loads the parsers legitimately peaks well above any ballast a test would allocate. That assertion measured "is the run small" when the property is "does the parent's history leak in". Rewritten to assert invariance across a released 192 MiB parent allocation.

Verified by mutation rather than by it passing: forcing the in-process path fails the test, and measured directly, the child peaks at 459 MiB with and without the ballast — identical, which is the property.

@greptileai review

**Comment by vitali87** [comment]
Fixed in f014e5ab. Correct again, and this one is the sharpest of the three: my previous commit fixed the module for Windows and left the test contradicting it.

`test_measure_indexing_records_wall_clock_and_memory` asserted `peak_rss_bytes > 0` unconditionally, while the same commit made that value `RSS_UNAVAILABLE` (-1) wherever `resource` is missing. The Windows fix could not pass its own suite.

**No local run could have caught it.** macOS has `resource`, so the sentinel branch never executes here and the assertion was true of every environment I can reach. The only signal was a platform I cannot run, which is exactly what the Windows CI job is for.

Two changes rather than the one suggested, because admitting the sentinel *removes* a check:

1. The assertion now accepts `RSS_UNAVAILABLE` while still rejecting `0` — what must never appear is a value that reads as a measurement without being one.
2. A new test pins the consequence: the sentinel must render `n/a`, never `-1 MiB` and never `0 MiB`. Widening a contract without pinning what the new state must look like would have traded a false failure for an unchecked path.

Also: `test_peak_memory_is_local_to_the_run_not_the_process` now skips when the sentinel is present rather than comparing `-1` with `-1`, which would pass vacuously and report Windows coverage that does not exist.

Verified the Windows path by simulating an absent `resource` module — probe returns `-1`, row renders `n/a`:

```
| Indexing time — c, 1 files (cgr 0.0.0) | 1.0s | 1 nodes / 0 edges | n/a |
```

Worth recording that all three findings on this PR were mine and none were reachable from my own machine: a Unix-only import, a process-wide RSS high-water mark, and a test contradicting its own implementation's platform branch.

@greptileai review

**Comment by vitali87** [comment]
@greptileai review

**Comment by vitali87** [comment]
Cleared the SonarCloud finding in `8c3066dc`.

One MAJOR code smell, queried from the API rather than inferred from the gate:

```
test_indexing_benchmark.py:136  Split this composite assertion into separate assertions.
```

That was `assert row.startswith("|") and row.endswith("|")`. Splitting it is a genuine improvement rather than gate appeasement: a composite failure cannot say which end is malformed, and **a row missing only its trailing pipe renders as broken markdown while the leading-pipe check still passes**. Each half now carries the row in its message.

Also checked the Scala tests on the sibling branch (#1472) for the same pattern — none, so this was the only instance.

@greptileai review

File: benchmarks/bench_indexing.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    `VmHWM` is exec-scoped on Linux; `ru_maxrss` is the fallback for platforms
+    with no `/proc` (macOS), where fork/exec does reset the high-water mark and
+    so the value is already child-local.
+    """
+    from_proc = _proc_peak_rss_bytes()
+    if from_proc is not None:
+        return from_proc
+    if resource is None:
+        return RSS_UNAVAILABLE
+    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss * _RSS_SCALE
```
[the remark this answers is not in the document]
Comment (vitali87) [reply benchmarks/bench_indexing.py]: Fixed in `f03797dc`. Correct, and it was the same bug hiding in the path I left behind: the `/proc` reader fixed the main path while the fallback quietly reinstated it, because `ru_maxrss` survives fork/exec on Linux.

On Linux an unreadable `/proc` now reports `RSS_UNAVAILABLE` rather than falling back. macOS keeps the fallback, where fork/exec does reset the high-water mark and the value is genuinely child-local.

Two paired tests, both faking the platform since macOS cannot execute either branch. The pair matters: without the second, the guard could be satisfied by disabling the fallback everywhere, losing a real measurement on the platform where it works.

Verified by mutation - restoring the unconditional fallback fails the Linux test with `assert 250478592 == -1`, a quarter-gigabyte of inherited parent history standing in for the sentinel.

File: benchmarks/bench_indexing.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    if os.environ.get(_CHILD_ENV_FLAG):
+        # Already the child: measure directly, and let the peak be this
+        # process's, which is exactly the run's.
+        return _measure_in_this_process(corpus, project_name)
+
+    completed = subprocess.run(
+        [
+            sys.executable,
+            "-m",
+            __spec__.name,
```
[the remark this answers is not in the document]
Comment (vitali87) [reply benchmarks/bench_indexing.py]: Fixed in `3595579d`, and reproduced first:

```
$ python benchmarks/bench_indexing.py /tmp/dcorp --project-name d
    __spec__.name,
    ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'name'
```

Running a file directly sets `__spec__` to None, so the relaunch died before the child was ever spawned. The benchmark failed on its most obvious invocation while working fine under `-m` - which is precisely why nothing caught it: every local run and every CI run used the module form.

`_module_name()` now falls back to the literal when `__spec__` is absent. Direct execution emits a real row:

```
| Indexing time - dcorp, 1 files (cgr 0.0.770) | 0.1s | 4 nodes / 3 edges | 81 MiB |
```

Two paired tests. The second asserts the real spec name WINS over the fallback under `-m`, so the fix cannot be implemented by always returning the hardcoded string - that would pass today and break silently the moment the module moves.

On the ast-grep CWE-78 note: the argv is built from `sys.executable`, a module-name literal, and the corpus path the operator passed in, with no shell. There is no request-sourced input here - it is a developer benchmark invoked from a terminal.

**Comment by vitali87** [comment]
@greptileai review

All other checks pass; this PR is waiting only on the 5/5 gate for head `3595579dd44e`.
